### PR TITLE
Manage PVCs in the conductor StatefulSet

### PR DESCRIPTION
**Pull request by steveb**
The intent is for each conductor pod to have its own volume to store
/var/lib/ironic state. However currently a single PVC is created and the
resulting volume is shared across conductors. This change moves the PVC
creation to be managed by the StatefulSet, so that PVC replicas match
pod replicas.

ReadWriteMany is not required for these volumes, and we required the
control plane to not provide volumes of this type, so this change also
switches to ReadWriteOnce. This fixes [OSPRH-7688](https://issues.redhat.com//browse/OSPRH-7688).

Jira: [OSPRH-7688](https://issues.redhat.com//browse/OSPRH-7688)

### CODE REVIEW

File: pkg/ironicconductor/pvc.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ func Pvc(instance *ironicv1.IronicConductor) *corev1.PersistentVolumeClaim {
 		},
 		Spec: corev1.PersistentVolumeClaimSpec{
 			AccessModes: []corev1.PersistentVolumeAccessMode{
-				corev1.ReadWriteMany,
+				corev1.ReadWriteOnce,
```
Comment (hjensas) [inline pkg/ironicconductor/pvc.go:35]: How will this work when we scale conductors to 2-3 replicas and the Pods are not scheduled on the same node?

```
ReadWriteOnce
    the volume can be mounted as read-write by a single node. ReadWriteOnce access mode still can allow multiple pods to access the volume when the pods are running on the same node. For single pod access, please see ReadWriteOncePod.
```

Don't we need to do something around here[[1]](https://github.com/openstack-k8s-operators/ironic-operator/blob/main/controllers/ironicconductor_controller.go#L189-L192) to ensure it creates a PVC for each replica in the statefulset?


[1] https://github.com/openstack-k8s-operators/ironic-operator/blob/main/controllers/ironicconductor_controller.go#L189-L192

File: pkg/ironicconductor/volumes.go [hunk 2]
```diff
@@ -13,9 +12,6 @@ import (
 func GetVolumes(instance *ironicv1.IronicConductor) []corev1.Volume {
 	var config0640AccessMode int32 = 0640
 	pvcName := fmt.Sprintf("%s-%s", ironic.ServiceName, ironic.ConductorComponent)
```
Comment (hjensas) [inline pkg/ironicconductor/volumes.go:14]: Something is not quite lined up... 

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-06-18T08:13:28Z"
    message: '0/3 nodes are available: persistentvolumeclaim "ironic-conductor" not
      found. preemption: 0/3 nodes are available: 3 Preemption is not helpful for
      scheduling..'
```

```
[zuul@controller-0 ~]$ oc get pvc ironic-conductor-ironic-conductor-0 
NAME                                  STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS    AGE
ironic-conductor-ironic-conductor-0   Pending                                      local-storage   34m
```

Comment (hjensas) [reply pkg/ironicconductor/volumes.go:14]: I think I figured this out. 
Not sure how to update this PR, so I created a new one: https://github.com/openstack-k8s-operators/ironic-operator/pull/439

The result looks like this after scaling from 1 to 3 replicas:
```
var-lib-ironic-ironic-conductor-0                          Bound     local-storage10-master-1   10Gi       RWO,ROX,RWX    local-storage   5m16s
var-lib-ironic-ironic-conductor-1                          Bound     local-storage03-master-2   10Gi       RWO,ROX,RWX    local-storage   35s
var-lib-ironic-ironic-conductor-2                          Bound     local-storage03-master-0   10Gi       RWO,ROX,RWX    local-storage   4s
```

```
[zuul@controller-0 ~]$ oc get ironicconductors.ironic.openstack.org ironic-conductor 
NAME               NETWORKATTACHMENTS   STATUS   MESSAGE
ironic-conductor   ["ironic"]           True     Setup complete
```

```
[zuul@controller-0 ~]$ oc get pod ironic-conductor-0 -o json | jq .spec.volumes[0]
{
  "name": "var-lib-ironic",
  "persistentVolumeClaim": {
    "claimName": "var-lib-ironic-ironic-conductor-0"
  }
}
[zuul@controller-0 ~]$ oc get pod ironic-conductor-1 -o json | jq .spec.volumes[0]
{
  "name": "var-lib-ironic",
  "persistentVolumeClaim": {
    "claimName": "var-lib-ironic-ironic-conductor-1"
  }
}
[zuul@controller-0 ~]$ oc get pod ironic-conductor-2 -o json | jq .spec.volumes[0]
{
  "name": "var-lib-ironic",
  "persistentVolumeClaim": {
    "claimName": "var-lib-ironic-ironic-conductor-2"
  }
}
```

volumeMounts for containers has:
```
      {
        "mountPath": "/var/lib/ironic",
        "name": "var-lib-ironic"
      },
```

Comment (steveb) [reply pkg/ironicconductor/volumes.go]: I pulled in the changes in https://github.com/openstack-k8s-operators/ironic-operator/pull/439, thanks

**Review by juliakreger — COMMENTED** [review]
/approve

**Pull request merged**